lsp: Add support for window/showDocument notification

imumesh18 opened the pull request:
Closes #17931
Closes #24852

| Before | After |
|--------|-------|
| <video src="https://github.com/user-attachments/assets/ef9dc782-21df-4a56-9774-cab62d20fa12" controls width="400"></video> | <video src="https://github.com/user-attachments/assets/2be6ef46-9dc4-4305-9a9d-72614e9c9f4d" controls width="400"></video> |

- [x] Handle window/showDocument to open both file and url non remote scenarios
- [x] Add tests for all the scenarios for remote and non-remote(suppress and show).
<strike>
- [x] Manual test that in collab it doesn't work at all.
- [ ] Manual test for remote
</strike>

cc @probably-neb Saw it got assigned to you. Just in case you like my implementation and would like to go with this approach would be more than happy work on this and get this through the line.

Release Notes:

- Add support for window/showDocument notification

imumesh18 committed:
lsp: Add support for window/showDocument notification

probably-neb commented:
This is great! Thanks for jumping on it so fast!

imumesh18 committed:
Add support for remote and collab usecases

imumesh18 committed:
oops 📎

imumesh18 committed:
cleanup and reuse existing proto message types

imumesh18 committed:
Update URL opening logic to use cx.open_url method

imumesh18 committed:
Merge branch 'main' into push-wrrouxqswmyx

imumesh18 committed:
Remove unused open dependency from project crate

osiewicz commented:
Nit-picking a bit, but does `cx.open_url` handle file paths for us? As in, I imagine that if the URI is a file path, we'd want to open a file in Zed (in the same window if it's a part of the same workspace, if possible).

osiewicz commented on crates/project/src/project_tests.rs:
```diff
@@ -9090,3 +9090,74 @@ async fn test_find_project_path_abs(
+            take_focus: Some(true),
+            selection: Some(lsp::Range::new(
+                lsp::Position::new(0, 0),
+                lsp::Position::new(0, 8),
```
You can prolly also check whether the selection is handled correctly.

osiewicz commented on crates/project/src/project_tests.rs:
```diff
@@ -9090,3 +9090,74 @@ async fn test_find_project_path_abs(
+        .unwrap();
+
+    second_buffer.read_with(cx, |buffer, _| {
+        assert_eq!(buffer.text(), "fn other() {}");
```
This assertion does not really do anything for us , as we're simply checking whether the buffer content is the same as it was when we've set up this test. Selections would be a good test (tho for that you might want to move this test into `editor_tests.rs`).

osiewicz commented on crates/project/src/project.rs:
```diff
@@ -4549,6 +4619,27 @@ impl Project {
+        mut cx: AsyncApp,
+    ) -> Result<()> {
+        this.update(&mut cx, |this, cx| {
+            if let Ok(uri) = lsp::Url::parse(&envelope.payload.uri) {
```
Should we be ignoring this error in particular? Erroring out on invalid URIs seems reasonable to me; we can still ignore non-file URIs, as there's no way for remote counterpart to handle any other URIs anyways.

osiewicz commented on crates/proto/proto/lsp.proto:
```diff
@@ -543,6 +543,14 @@ message UpdateLanguageServer {
+message ShowDocument {
+    uint64 project_id = 1;
+    string uri = 2;
+    bool take_focus = 3;
+    optional PointUtf16 selection_start = 4;
+    optional PointUtf16 selection_end = 5;
```
I feel like `take_focus` and `selection_start`/`selection_end` don't make sense in RPC calls, as they can only be handled (reasonably) by the requesting side.

We should prolly return buffer_id from `ShowDocument` RPC call.

osiewicz commented on crates/project/src/lsp_store.rs:
```diff
@@ -1012,6 +1012,40 @@ impl LocalLspStore {
+                                    take_focus: params.take_focus.unwrap_or(true),
+                                    selection: params.selection,
+                                });
+                                true
+                            })
+                            .unwrap_or(false);
```
```suggestion
                            }).is_ok();
```

imumesh18 committed:
Merge branch 'main' into push-wrrouxqswmyx

imumesh18 committed:
Suppress ShowDocument in collab and refactor request handling

- Suppress LSP ShowDocument requests in collaborative scenarios to avoid
disruptive behavior for all participants - Refactor ShowDocument to be a
request/response instead of a message - Remove unused fields (take_focus,
selection) from proto::ShowDocument - Update tests for ShowDocument behavior in
local, remote, and collab modes

imumesh18 committed:
suppress ShowDocument requests in collaborative scenarios

imumesh18 committed:
code cleanup

imumesh18 commented:
Okay so i have completely disabled it for remote. First I thought except for collab it should work for atleast remote but it's the problem I was facing while writing test as well manually there is currently no way to figure out and differentiate between(atleast I couldn't find it) on how can we make it work just for remote and not for collab . Even I tried to see if we can disable it for collab clients and have it for host(couldn't figured it out) and this lead to issue in remote project because doing this we need to figure out if it's a remote project client who is requesting it or a collab client which something again I couldn't see if it was possible. Just wanted to really understand what @osiewicz really meant when he mentioned in his comments that this will really difficult to implement in remote setup which turned out to be true. So at the end it did make sense to just scrap remote all together. I couldn't figure out if there is a work around by making some changes in the way this currently handled so just made sure we bail in case of remote and handle in case of local by showing the doc.

osiewicz commented:
@imumesh18 I think we can implement it on the remote counterpart, as long as there's not a single collaborator. Either way, I'll take a review pass tomorrow.

imumesh18 committed:
revert back to just making it work with local

imumesh18 committed:
Merge branch 'main' into push-wrrouxqswmyx

imumesh18 commented:
So the current types of urls are following there are three types of url which will be sent from lsp

* Public url like github.com or doc.go.dev etc.
* Doc served by lsp server like 127.0.0.1:1221/some-package/some-struct/some-doc.
* File url which is nothing but some local file.

I believe we can implement it like following for different scenarios of zed:

* Local: Support all the type of url.
* Remote: Can support file url and public url. Theoretically this should work for both these cases but still need to verify if cx.open_url will trigger the browser on host or remote. Might come across as a broken implementation.
* Collab: Should skip altogether.

Let me know your thought on this.

osiewicz commented:
> Theoretically this should work for both these cases but still need to verify if cx.open_url will trigger the browser on host or remote
`cx.open_url` will trigger on remote host (machine running your project), so we need to transfer that over RPC. 
> Though will need to figure out how to distinguish it from collab.
There's a set of functions you can use to check:
- a list of collaborators: https://github.com/zed-industries/zed/blob/afab4b522e5b4ee9d9b48a4cb473a2b060530eba/crates/project/src/project.rs#L1853 (this may include the "owner" [0] of a project and it's remote host[1], not sure)
- other bits about the project: https://github.com/zed-industries/zed/blob/afab4b522e5b4ee9d9b48a4cb473a2b060530eba/crates/project/src/project.rs#L2355-L2378 `is_via_collab` will return `false` for project owners and remote hosts, so you most likely don't need to care about it. 

I imagine it will look more or less like:
```
def on_show_document_request_from_language_server(...):
  if self.is_local():
    cx.open_url(params.url)
  elif self.is_via_ssh():
    self.send_rpc_message(self.host()?) # Send rpc message ONLY to the project owner and nobody else.
  else:
    ## this is is_collab branch. It is effectively unreachable, because `on_show_document` lsp notification can only be received by whoever is running the language server.

def on_show_document_rpc_message(...):
  # Ran on project owner when the remote host sends a rpc message
  cx.open_url(params.url)

```


[0]: A user that opened the project via SSH. That collaborator is a project owner.
[1]: A "fake" participant that runs on the remote machine. This is not a human, though it is "owned" by project owner.

imumesh18 committed:
implement showDocument for remote and skip for collab

imumesh18 committed:
initial tests

imumesh18 committed:
Merge branch 'main' into push-wrrouxqswmyx

imumesh18 commented:
Hey @probably-neb need help. So I was trying to test this on remote development server. The problem is I am not able to build the remote development server on my local ssh container. So here is my setup I run `ZED_BUILD_REMOTE_SERVER="cross" cargo run` and then from the dev build I ssh into my docker container which works fine with other releases but not when I do with dev release it starts to build the remote server and fails with following error. Not been able to figure out anything. If you have any pointers how can I make it work It will be much appreciated.

<details>
<summary>Logs</summary>

```
rror: linking with `aarch64-linux-gnu-gcc` failed: exit status: 1
  |
  = note:  "aarch64-linux-gnu-gcc" "-Wl,--version-script=/tmp/rustc8Z8gdB/list" "-Wl,--no-undefined-version" "/tmp/rustc8Z8gdB/symbols.o" "<17 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/target/aarch64-unknown-linux-gnu/debug/deps/{librustls_native_certs-97da4f8ac4ba6500.rlib,libopenssl_probe-ec100fa02286de19.rlib,libonce_cell-531e87b3250bd46a.rlib,librustls-c86dc7be61bb4adb.rlib,libsubtle-4bb4224fd8f524e3.rlib,libwebpki-92e8c1f4ce553ac7.rlib,libaws_lc_rs-6d654e5ff2489fc2.rlib,libaws_lc_sys-210f85b0997ef427.rlib,libring-2881448eec1573d2.rlib,libgetrandom-0804afdf94225efe.rlib,libuntrusted-d57b0444b67bd026.rlib,liblibc-1e60937ace3ff966.rlib,libcfg_if-173222d37a466d46.rlib,librustls_pki_types-cc55b3af7f44efde.rlib,libzeroize-390388b5d5bf3bf0.rlib,liblog-a3b0e580554b3250.rlib,libvalue_bag-3a7180f8ef3e5c02.rlib,libvalue_bag_serde1-0df26082345ebd22.rlib,libserde_fmt-aeb3fb8e3a8fb9c8.rlib,liberased_serde-2403b54192785bae.rlib,libtypeid-9616cd765f59fb8f.rlib,libserde-d9b33fb0f81cac4a.rlib}.rlib" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustc8Z8gdB/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/target/aarch64-unknown-linux-gnu/debug/build/aws-lc-sys-b84ce669fd22c163/out" "-L" "/target/aarch64-unknown-linux-gnu/debug/build/ring-a90af89cb5165e1c/out" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/target/aarch64-unknown-linux-gnu/debug/deps/librustls_platform_verifier-142ce92be2b8fcb0.so" "-Wl,--gc-sections" "-shared" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-fuse-ld=mold"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: aarch64-linux-gnu-gcc: error: unrecognized command line option '-fuse-ld=mold'; did you mean '-fuse-ld=gold'?


error: could not compile `rustls-platform-verifier` (lib) due to 1 previous error
```

</details>

imumesh18 commented:
Thanks @probably-neb, it seems to be an issue with cross base image args. Okay so to update you about this. I have tested all the following scenarios except as collab guest. All of them are working might need your help to simulate the real scenario for collab guest to see it works as expected for both scenarios. Only thing remaining is little bit of code cleanup around is_local_url and test cleanup. This should be good for review till then gonna keep it in draft. All thanks to @osiewicz for pointing me in right direction for collab and remote part. We have a base line for v1 how this should work atleast for now.

| Project Type | File URL | Public URL (https://) | Localhost URL (http://localhost) |
|--------------|----------|----------------------|----------------------------------|
| **Local** | ✅ Opens file in buffer | ✅ Opens in browser | ✅ Opens in browser |
| **Remote** | ✅ Opens file in buffer (forwarded to host) | ✅ Opens in browser (forwarded to host) | ❌ Not supported |
| **Collab Host** | ✅ Opens file in buffer (LSP runs locally) | ✅ Opens in browser | ✅ Opens in browser |
| **Collab Guest** | ❌ Cannot trigger ShowDocument | ❌ Cannot trigger ShowDocument | ❌ Cannot trigger ShowDocument |
| **SSH Remote** | ✅ Opens file in buffer (forwarded to SSH host) | ✅ Opens in browser (forwarded to SSH host) | ❌ Not supported |
| **SSH + Collab Host** | ✅ Opens file in buffer (LSP on remote) | ✅ Opens in browser (forwarded to SSH host) | ❌ Not supported |
| **SSH + Collab Guest** | ❌ Cannot trigger ShowDocument | ❌ Cannot trigger ShowDocument | ❌ Cannot trigger ShowDocument |

imumesh18 committed:
Skip opening localhost URLs in show document handler

imumesh18 committed:
Merge branch 'main' into push-wrrouxqswmyx

imumesh18 committed:
📎

imumesh18 commented:
Going to close this because of too many merge conflict and didn't get much time to revisit this.
